From 04859b72fae4d60ad02e4d90bc25329a9111b45f Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Sat, 18 Mar 2023 09:30:08 +0000
Subject: [PATCH 04/53] [Huawei] mm/sharepool: Get sp_group from mm if possible

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

This could skip to check if current is added to the sp_group because the
group is found from the task's group list.

sp_alloc, sp_free, make_share_k2u and unshare_uva are related.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 119 +++++++++++++++++++++++-------------------------
 1 file changed, 57 insertions(+), 62 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index a3ae16f06fd6..a74e09f98fd5 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -666,8 +666,6 @@ static struct sp_overall_stat sp_overall_stat;
 
 enum spa_type {
 	SPA_TYPE_ALLOC = 1,
-	/* NOTE: reorganize after the statisical structure is reconstructed. */
-	SPA_TYPE_ALLOC_PRIVATE = SPA_TYPE_ALLOC,
 	SPA_TYPE_K2TASK,
 	SPA_TYPE_K2SPG,
 };
@@ -961,6 +959,37 @@ static bool is_process_in_group(struct sp_group *spg,
 	return false;
 }
 
+static struct sp_group *sp_group_get_from_mm(struct mm_struct *mm, int spg_id)
+{
+	struct sp_group *spg = NULL;
+	struct sp_group_node *spg_node;
+	struct sp_group_master *master;
+
+	down_read(&sp_group_sem);
+
+	master = mm->sp_group_master;
+	if (!master) {
+		up_read(&sp_group_sem);
+		return NULL;
+	}
+
+	if (spg_id == SPG_ID_DEFAULT) {
+		atomic_inc(&master->local->use_count);
+		up_read(&sp_group_sem);
+		return master->local;
+	}
+
+	list_for_each_entry(spg_node, &master->group_head, group_node)
+		if (spg_node->spg->id == spg_id) {
+			if (atomic_inc_not_zero(&spg_node->spg->use_count))
+				spg = spg_node->spg;
+			break;
+		}
+	up_read(&sp_group_sem);
+
+	return spg;
+}
+
 /* user must call sp_group_put() after use */
 static struct sp_group *sp_group_get_locked(int tgid, int spg_id)
 {
@@ -1970,18 +1999,6 @@ static void sp_free_unmap_fallocate(struct sp_area *spa)
 	up_read(&spa->spg->rw_lock);
 }
 
-static int sp_check_caller_permission(struct sp_group *spg, struct mm_struct *mm)
-{
-	int ret = 0;
-
-	down_read(&spg->rw_lock);
-	if (!is_process_in_group(spg, mm))
-		ret = -EPERM;
-	up_read(&spg->rw_lock);
-
-	return ret;
-}
-
 #define FREE_CONT	1
 #define FREE_END	2
 
@@ -2000,9 +2017,16 @@ static int sp_free_get_spa(struct sp_free_context *fc)
 	struct sp_area *spa;
 	struct sp_group *spg;
 
-	spg = sp_group_get(current->tgid, fc->spg_id);
+	if (!current->mm) {
+		/*
+		 * XXXX: should process this branch carefully instead of return an error.
+		 */
+		return -EINVAL;
+	}
+
+	spg = sp_group_get_from_mm(current->mm, fc->spg_id);
 	if (!spg) {
-		pr_debug("sp free get group failed %d\n", fc->spg_id);
+		pr_err("sp free get group failed %d\n", fc->spg_id);
 		return -EINVAL;
 	}
 
@@ -2022,19 +2046,6 @@ static int sp_free_get_spa(struct sp_free_context *fc)
 	}
 	fc->spa = spa;
 
-	if (!current->mm)
-		goto check_spa;
-
-	ret = sp_check_caller_permission(spa->spg, current->mm);
-	if (ret < 0)
-		goto drop_spa;
-
-check_spa:
-	if (is_local_group(spa->spg->id) && (current->tgid != spa->applier)) {
-		ret = -EPERM;
-		goto drop_spa;
-	}
-
 	down_write(&spa->spg->rw_lock);
 	if (!spg_valid(spa->spg)) {
 		fc->state = FREE_END;
@@ -2181,9 +2192,9 @@ static int sp_alloc_prepare(unsigned long size, unsigned long sp_flags,
 		sp_flags |= SP_HUGEPAGE;
 
 	if (spg_id != SPG_ID_DEFAULT) {
-		spg = sp_group_get(current->tgid, spg_id);
+		spg = sp_group_get_from_mm(current->mm, spg_id);
 		if (!spg) {
-			pr_err_ratelimited("allocation failed, can't find group\n");
+			pr_err("allocation failed, can't find group(%d)\n", spg_id);
 			return -ENODEV;
 		}
 
@@ -2195,22 +2206,14 @@ static int sp_alloc_prepare(unsigned long size, unsigned long sp_flags,
 			pr_err_ratelimited("allocation failed, spg is dead\n");
 			return -ENODEV;
 		}
-
-		if (!is_process_in_group(spg, current->mm)) {
-			up_read(&spg->rw_lock);
-			sp_group_put(spg);
-			pr_err_ratelimited("allocation failed, task not in group\n");
-			return -ENODEV;
-		}
-		ac->type = SPA_TYPE_ALLOC;
 	} else {  /* allocation pass through scene */
 		spg = sp_get_local_group(current, current->mm);
 		if (IS_ERR(spg))
 			return PTR_ERR(spg);
 		down_read(&spg->rw_lock);
-		ac->type = SPA_TYPE_ALLOC_PRIVATE;
 	}
 
+	ac->type = SPA_TYPE_ALLOC;
 	if (sp_flags & SP_HUGEPAGE) {
 		ac->file = spg->file_hugetlb;
 		ac->size_aligned = ALIGN(size, PMD_SIZE);
@@ -2789,14 +2792,8 @@ void *mg_sp_make_share_k2u(unsigned long kva, unsigned long size,
 		spg_id = SPG_ID_DEFAULT;
 	}
 
-	spg = sp_group_get(current->tgid, spg_id);
+	spg = sp_group_get_from_mm(current->mm, spg_id);
 	if (spg) {
-		ret = sp_check_caller_permission(spg, current->mm);
-		if (ret < 0) {
-			sp_group_put(spg);
-			uva = ERR_PTR(ret);
-			goto out;
-		}
 		uva = sp_make_share_kva_to_spg(&kc, spg);
 		sp_group_put(spg);
 	} else {
@@ -3123,8 +3120,8 @@ EXPORT_SYMBOL_GPL(mg_sp_make_share_u2k);
  * sp_unshare_uva - unshare a uva from sp_make_share_k2u
  * @uva: the uva to be unshared
  * @size: not used actually and we just check it
- * @group_id: specify the spg of the uva; for local group, it can be SPG_ID_DEFAULT
- *            unless current process is exiting.
+ * @spg_id: specify the spg of the uva; for local group, it can be SPG_ID_DEFAULT
+ *          unless current process is exiting.
  *
  * Procedure of unshare uva must be compatible with:
  *
@@ -3133,16 +3130,24 @@ EXPORT_SYMBOL_GPL(mg_sp_make_share_u2k);
  * exit_task_work() -> task_work_run() -> __fput() -> ... -> vdec_close() ->
  * sp_unshare(uva, local_spg_id)
  */
-static int sp_unshare_uva(unsigned long uva, unsigned long size, int group_id)
+static int sp_unshare_uva(unsigned long uva, unsigned long size, int spg_id)
 {
 	int ret = 0;
 	struct sp_area *spa;
 	unsigned int page_size;
 	struct sp_group *spg;
 
-	spg = sp_group_get(current->tgid, group_id);
+	if (!current->mm) {
+		/*
+		 * XXXX: should process this branch carefully instead of return an error.
+		 */
+		return -EINVAL;
+	}
+
+	spg_id = (spg_id == SPG_ID_NONE) ? SPG_ID_DEFAULT : spg_id;
+	spg = sp_group_get_from_mm(current->mm, spg_id);
 	if (!spg) {
-		pr_err("sp unshare find group failed %d\n", group_id);
+		pr_err("sp free get group failed %d\n", spg_id);
 		return -EINVAL;
 	}
 
@@ -3173,16 +3178,6 @@ static int sp_unshare_uva(unsigned long uva, unsigned long size, int group_id)
 		goto out_drop_area;
 	}
 
-	down_read(&spa->spg->rw_lock);
-	/* always allow dvpp channel destroy procedure */
-	if (current->mm && !is_process_in_group(spa->spg, current->mm)) {
-		up_read(&spa->spg->rw_lock);
-		pr_err("unshare uva failed, caller process doesn't belong to target group\n");
-		ret = -EPERM;
-		goto out_drop_area;
-	}
-	up_read(&spa->spg->rw_lock);
-
 	down_write(&spa->spg->rw_lock);
 	if (!spg_valid(spa->spg)) {
 		up_write(&spa->spg->rw_lock);
-- 
2.17.1

